cpp/src/join/semi_join.cu

```diff
@@ -77,90 +296,33 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
-
-  auto right_flattened_keys = right_flattened_tables.flattened_columns();
-  auto left_flattened_keys  = left_flattened_tables.flattened_columns();
-
-  // Create hash table.
-  semi_map_type hash_table{compute_hash_table_size(right_num_rows),
-                           cuco::sentinel::empty_key{std::numeric_limits<hash_value_type>::max()},
-                           cuco::sentinel::empty_value{cudf::detail::JoinNoneValue},
-                           hash_table_allocator_type{default_allocator<char>{}, stream},
-                           stream.value()};
-
-  // Create hash table containing all keys found in right table
-  auto right_rows_d      = table_device_view::create(right_flattened_keys, stream);
-  auto const right_nulls = cudf::nullate::DYNAMIC{cudf::has_nulls(right_flattened_keys)};
-  row_hash const hash_build{right_nulls, *right_rows_d};
-  row_equality equality_build{right_nulls, *right_rows_d, *right_rows_d, compare_nulls};
-  make_pair_fn pair_func_build{};
-
-  auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func_build);
-
-  // skip rows that are null here.
-  if ((compare_nulls == null_equality::EQUAL) or (not nullable(right_keys))) {
-    hash_table.insert(iter, iter + right_num_rows, hash_build, equality_build, stream.value());
-  } else {
-    thrust::counting_iterator<size_type> stencil(0);
-    auto const [row_bitmask, _] = cudf::detail::bitmask_and(right_flattened_keys, stream);
-    row_is_valid pred{static_cast<bitmask_type const*>(row_bitmask.data())};
-
-    // insert valid rows
-    hash_table.insert_if(
-      iter, iter + right_num_rows, stencil, pred, hash_build, equality_build, stream.value());
-  }
-
-  // Now we have a hash table, we need to iterate over the rows of the left table
-  // and check to see if they are contained in the hash table
-  auto left_rows_d      = table_device_view::create(left_flattened_keys, stream);
-  auto const left_nulls = cudf::nullate::DYNAMIC{cudf::has_nulls(left_flattened_keys)};
-  row_hash hash_probe{left_nulls, *left_rows_d};
-  // Note: This equality comparator violates symmetry of equality and is
-  // therefore relying on the implementation detail of the order in which its
-  // operator is invoked. If cuco makes no promises about the order of
-  // invocation this seems a bit unsafe.
-  row_equality equality_probe{left_nulls, *right_rows_d, *left_rows_d, compare_nulls};
-
-  // For semi join we want contains to be true, for anti join we want contains to be false
-  bool const join_type_boolean = (kind == join_kind::LEFT_SEMI_JOIN);
-
-  auto hash_table_view = hash_table.get_device_view();
-
+  auto const flagged = [&] {
+    // Use `nan_equality::UNEQUAL` as the default value for comparing NaNs in semi- and anti- joins.
+    auto contained =
+      left_semi_join_contains(left_keys, right_keys, compare_nulls, nan_equality::UNEQUAL, stream);
+    if (kind == join_kind::LEFT_ANTI_JOIN) {
```

Comment:
Rather than launching this extra kernel, we should fuse it into the `copy_if` kernel using a custom comparator that negates the condition for a `LEFT_ANTI_JOIN`.

Reply:
>  we should use the bullk API from cuco

To clarify, my intention is to use bulk APIs whenever applicable i.e. avoid code duplication and not hurt runtime performance. If an extra kernel is required to incorporate the bulk API, it may not be a proper use case.

Reply:
I am personally curious to see what the result would be of using the device APIs for the map and the quality comparator to define a custom predicate for `copy_if`. There is the serious downside that this approach would result in significant code duplication since we wouldn't be reusing the new `contains` API. However, I suspect that not materializing the contains result and converting three kernel launches (`contains`, `thrust::transform`, and `thrust::copy_if`) into just one (`thrust::copy_if` with a custom predicate) will give substantial enough performance benefits to maybe be worthwhile. This sort of goes against the original purpose of this PR, though, which was just to factor this `contains` logic out of the `semi_anti_join` code so that it could be reused elsewhere, so I don't think we need to try it now. We can make a note of it as a potential future optimization.

Reply:
I remember that there was a PR doing the opposite thing. This one: https://github.com/rapidsai/cudf/pull/10511

So unifying `copy_if` with `contains` in fact makes the performance worse :|

Reply:
Oh this was that change. I did not make that connection. So I guess I'm not the only one to make this mistake before :sweat_smile: It seems that the register usage overhead incurred by the hash map lookups is super high. No need to take this discussion any further, then.

Reply:
@ttnghia actually, could you add a comment in this file indicating _why_ we materialize the gather map? The lack of a comment means that someone else will almost certainly think like me and try to convert this back in the future!

Reply:
This is a good point. I'll do it.